Add ADR proposal for "requirements for SSO identity federation" (#205)
SovereignCloudStack/standards: Decisions/scs-XXXX-v1-requirements-for-sso-identity-federation.md

users, it would be good to be able to consume these external identities in SCS.

For customers this avoids the neccessity to explicitly maintain an additional
dedicated account in SCS and this also reduces what SCS needs to do with
respect to handling storage and lifecycle maintenance for user accounts.

To put it in other workds, in SCS we would like to be able to delegate
authentication to external identity providers and map those users to roles in
SCS that can be used for authorization decisions when users access SCS services.

# Motivation for this document

SCS has multiple service layers, like IaaS and CaaS, both of which running their
own technological stack with specific internal models of accounts and
authorization. OTOH there are also specific services like the "Status Page"
application that require consumption of identites.

One thing these services have in common, is that they are able
to use SSO protocols like OAuth 2.0 or OpenID Connect (OIDC) on top of it to
delegate authentication. They are service providers (SAML terminology) and can
be relying parties (OIDC terminology) of a protocol compliant identity provider
(IdP).

> So the idea is, to run an SSO IdP as part of SCS to provide a dedicated point

Author: I was told by an english speaker (better than me) that it depends on the question if the following word is *spoken* with a starting vocal rather that if it *written* with a starting vocal or consonant. I hope my sentence makes sense.. But I don't mind and will accept your proposal.

> So the idea is, to run an SSO IdP as part of SCS to provide a dedicated point

Reviewer: Did not think about that. TIL. Also TIL that the English name of "S" is really "ess", according to Wikipedia.

of entry for identites, which the SCS service layers can use as a common

> interface to consume external user identities.

Reviewer: Does this mean that SCS does not offer to maintain (persisted) identities itself?
Or that there are multiple IdP's? One for SCS-local users and one for federated ones?

Author: This paragraph only talks about the topic of external identities (the document is about federation after all). But you are right, there probably is also a need for SCS-local identities. I've only mentioned it below `ephemeral workload identities`. Maybe we should also mention this here. I'm unsure, because I don't have a clear picture for this. We could theoretically, as you mention, have one IdP for federation and use a separate IdP for storage of actual SCS-local identities. As a physicist I like the symmetry. But I think that's overkill technically. I'd rather store them either inside of the IdP or in an attached LDAP compliant IAM backend.

Reviewer: +1

I think that it would be beneficial to explicitly state that maintenance of SCS-local identities is also not decided yet. Otherwise a reader might think a) that maintaining identities **within** SCS is done already and b) that the whole identity federation feature is an opt-in feature, as it is kind-of suggested above, by saying that only a subset of all customers will profit from it:

> This is based on the observation that prospective customers of a SCS based CSP
> **sometimes** already come equipped with an IAM solution of their choice

On-topic: I like the perspective of having a single IdP handling all identities - no matter where they come from (user identities primarily managed by the IdP itself; workload identities; user identities from 3rd party IdP's; etc.) as hinted here:

> So the idea is, to run a SSO IdP as part of SCS to provide a dedicated point
> of entry for identites,

Author: Good point, timely following up to today's discussion. Added in latest commit

The purpose of this document is to sepcify what requirements a specific
technical IdP implementation (i.e. software solution) needs to fulfill
in the context of SCS.

# Design Considerations

As a central service for identity handling, the IdP

> service needs to be realy really robust.

Reviewer: ```suggestion
service needs to be really, really robust.
```

Customers shall be able to access self service, so that
they can make reasonable adjustments e.g. to role mapping.
At the time of writing this document it's still undecided

> if SCS has the requirement of a decicated "self service"

Reviewer: ```suggestion
if SCS has the requirement of a dedicated "self service" service
```

The latter "service" may be redundant, but also may improve reading flow by differentiating between "self service" (a pattern) and "service" (a thing that can do things like serving a frontend). Maybe there is also better way to express that.

Author: I added and removed that additional "service" twice because I was unsure. I accept your proposal.

that serves as a frontend to provision and re-configure
customer specific data, abstracting e.g. from IdP specific
user interface particularities.

## Options considered

### Keycloak

Keycloak is a commonly used IdP solution implemented in Java.

> It is developped as an open source community project.

Reviewer: ```suggestion
It is developed as an open source community project.
```

as listed on the [community section](https://www.keycloak.org/community)
of the project website. It uses [Github issues](https://github.com/keycloak/keycloak/issues)
to track development.

It offers a REST API for administration and there's a separately maintained
3rd party python module as well as ansible support for it. Both of these are
downstream of Keycloak itself and may thus not always be feature complete and
suffer latency with respect to getting adjusted to upstream changes.

It offers support for commonly used SSO protocols and is "reasonably" fast
in adopting to protocol standard changes and extensions. This has been
observed in the case of logout support (backend and frontend variants) in OIDC.

It offers a concept of "Identity Brokering", where Keycloak is not just IdP
but also "client" to other IdPs. This allows daisy chaining of identity
federation. In this configuration it can work as a point of protocol
transition between different supported SSO protocols (SAML, OAuth 2.0, etc.).

Beyond this capability of using other IdPs as identity sources, it also supports
using classic LDAP based IAM services as backend (OpenLDAP and Active Directory,
e.g.).

> Keycloaks implementation makes some design decisions, that are specific

Reviewer: ```suggestion
Keycloak's implementation makes some design decisions, that are specific
```

to it and have consequences for clients of the service. E.g. Keycloak
has a concept of management "Realms", which have their own specific
set of HTTP API entrypoints, both for administration as well as for IdP
requests.

Commonly Keycloak realms can be used to map them 1:1 to user domains,
but since Keycloak supports configuring multiple backend IdPs in a
realm to be used for "Identity Brokering", there is always the
possibility to create a kind of "proxy" realm to provide a single
standard set of HTTP API endpoints for SSO clients (service providers)
to avoid the need to frequently extend/reduce client service configuration
whenever a new IdP federation needs to be added to Keycloak to onboard
a new customer. This is relevant for services like OpenStack Keystone,
which currently cannot be easily reconfigured for new SSO endpoints
without restarting the service, making the service unavailable for
a short span of time and increasing risk connected with service restarts.

Since version 17, Keycloak claims that it's capability for

> "cloud neative" deployments on Kubernetes has improved.

Reviewer: ```suggestion
"cloud native" deployments on Kubernetes has improved.
```

[Keycloak is offering a REST API](https://www.keycloak.org/docs-api/20.0.1/rest-api/index.html)
for all aspects of its administration interface.

For storage of Keycloak configuration and local user metadata
(e.g. from which external IdP a user account originally came from)
Keycloak supports several SQL backends through JDBC. Thus
it can be hooked up to a Postgres Database or to a
MariaDB/Galera cluster e.g..

> Keycloak is currently beeing deployed as part of the OSISM testbed.

Reviewer: ```suggestion
Keycloak is currently being deployed as part of the OSISM testbed.
```

Technically this IdP component shall be shifted from the management
plane to be run on the basis of a "minimal" Kubernetes (e.g. K3S),
e.g. to make use of the "self healing" and scaling features achievable
with that. As a central service for identity handling, the IdP

> service needs to be realy really robust.

Reviewer: This is a repetition of the same sentence above.

*Maybe* Kubernetes's features could also be omitted, as they are not too relevant.

Author: I think it is a relevant aspect for the selection of an IdP solution: Does it work well on a K8S environment or is it maybe not designed for that at all? The instances will need to share configuration (probably via the shared backend database) as well as session state. Maybe some are better prepared for horizontal scaling than others.

Reviewer: Agreed. Maybe just rephrase it exactly like that.
And also move it into the requirements/"design considerations" section or give Zitadel a matching paragraph as well?

Author: Good point, yeah, I was too much in "Brainstorm" mode when writing about the specific options. I'll adjust this.

Author: Adjusted in latest commit.

### Zitadel

> Zitadel is a newer implementation of an SSO IdP. It is implemented

Reviewer: ```suggestion
Zitadel is a newer implementation of a SSO IdP. It is implemented
```

in the following areas:
* For client services (single set of HTTP API endpoints).
* For SCS operators for provisioning customer [organizations](https://zitadel.com/docs/concepts/structure/organizations)
  and robust configuraton by using templated client, role and mapping
  configuration.
* For SCS customers for a robust user experience for self servicing.

[Zitadel is offering REST APIs](https://zitadel.com/docs/apis/introduction)
for multiple areas of use and configuration.

In comparison to Keycloak it currently lacks support for the
[Device Authorization Grant](https://github.com/SovereignCloudStack/issues/issues/221),
which, at time of writing, is a feauture that is relevant
for SCS to be able use OpenStack CLI and APIs with federated
identities. At Zitadel there are at least two [Github issues](https://github.com/zitadel/oidc/issues/141)
proposed to implement this feature.

Currently it still lacks support to consume identity data from LDAP backends.
This may become relevant in the context of SCS in case we would choose
an LDAP compliant service to store and retrieve SCS local accounts (or
account metadata) for example to have ephemeral workload identities in

> the CaaS layer or to track life cycle of SCS artefacts for federated identities.

Reviewer: How would LDAP relate to ephemeral workload identities here?

In general: Isn't Zitatel able to store local accounts itself?

Author: Sure, you can store stuff Zitadel (or rather in the cockroach DB backend) just like you can in Keycloak. It's just the question if it's their strong point to act as backend storage for identities. I've used very open wording here, as we simply don't know yet.

Reviewer: Ah, ok. Trying to reiterate what I understood (without ever having used Zitadel/Keycloak yet):

So comparing Zitadel to Keycloak, both allow to store users in their own database.
But doing so is not a strong point for Keycloak. For Keycloak, a common solution would be to rely on an LDAP server, whose whole point is storing users in its database.

So, if neither storing users in its own database, nor getting information from some e. g. LDAP server, is a strong feature in Zitadel, it would be (in that regard) inferior to Keycloak which at least happily integrates with some LDAP server.

Did I got that right?

Author: I'd say yes, but that may sound biased. I'd like to at least keep an open mind.

Author: The crucial point for Zitadel is the ease of integrating N+1 customer organization. At that point a single set of API endpoints (as opposed to one set of Endpoints per customer) may help avoiding reconfiguration issues in Keystone. I guess somebody would need to do a PoC to compare. I mean, I still have to communicate somehowbetween OpenStack/Keystone and Keycloak  to which organization I belong. I have not seen this in action yet.

Author: In the latest commit I've reduced the discussion about LDAP in the document.

track: IAM
---

# Introduction

Our assumption is that there are use cases, where CSPs would like to be able to
let customers access their SCS based services by identifying themselves with
credentials that are stored and managed external to the CSP's SCS installation.

This is based on the observation that prospective customers of a SCS based CSP
sometimes already come equipped with an IAM solution of their choice, either on
premises or e.g. as an external 3rd party cloud service. To ease onboarding of
customer employees (or e.g. customer contracted 3rd party admin staff) as SCS
users, it would be good to be able to consume these external identities in SCS.

For customers this avoids the neccessity to explicitly maintain an additional
dedicated account in SCS and this also reduces what SCS needs to do with
respect to taking care of persisting user account information.

To put it in other words, in SCS we would like to be able to delegate
authentication to external identity providers and map those users to roles in
SCS that can be used for authorization decisions when users access SCS services.

In addition to user identities there we also see the necessity to support the

> use of "machine identites" (aka "workload identities" or "service accounts).

Reviewer 2: ```suggestion
use of "machine identites" (aka "workload identities" or "service accounts”).
```

These will probably be SCS-local accounts and have for example the purpose
to grant CaaS workload access to storage resources served by the infrastructure
layer. Exact architectural details for this are still in active discussion,
but it is anticipated that the IdP component should be very useful in
facilitating the integration.

# Motivation for this document

SCS has multiple service layers, like IaaS and CaaS, both of which running their
own technological stack with specific internal models of accounts and

> authorization. OTOH there are also specific services like the "Status Page"

Reviewer 2: I’d actually not reference the Status Page here, since that should run elsewhere and not rely on anything that is part of the infrastructure for which it provides “Status Infos” (otherwise if the castle is burnt down it can’t display “All things broken” :)

application that require consumption of identities.

One thing these services have in common, is that they are able
to use SSO protocols like OAuth 2.0 or OpenID Connect (OIDC) on top of it to
delegate authentication. They are service providers (SAML terminology) and can
be relying parties (OIDC terminology) of a protocol compliant identity provider
(IdP).

So the idea is, to run a SSO IdP as part of SCS to provide a dedicated point
of entry for identites, which the SCS service layers can use as a common
interface to consume external user identities.

The purpose of this document is to specify what requirements a specific
technical IdP implementation (i.e. software solution) needs to fulfill
in the context of SCS.

# Design Considerations

> As a central service for identity handling, the IdP
> service needs to be really, really robust.

Reviewer 2: ```suggestion
As a central service for identity handling, the IdP
service needs to be robust and reliable.
```

Customers shall be able to access self service, so that
they can make reasonable adjustments e.g. to role mapping.
At the time of writing this document it's still undecided
if SCS has the requirement of a dedicated "self service" service
that serves as a frontend to provision and re-configure
customer specific data, abstracting e.g. from IdP specific
user interface particularities.

> Keycloak is currently being deployed as part of the OSISM testbed.

Reviewer 2: ```suggestion
Keycloak is currently being deployed as part of the IaaS reference implementation.
```

realm to be used for "Identity Brokering", there is always the
possibility to create a kind of "proxy" realm to provide a single
standard set of HTTP API endpoints for SSO clients (service providers)
to avoid the need to frequently extend/reduce client service configuration
whenever a new IdP federation needs to be added to Keycloak to onboard
a new customer. This is relevant for services like OpenStack Keystone,
which currently cannot be easily reconfigured for new SSO endpoints
without restarting the service, making the service unavailable for
a short span of time and increasing risk connected with service restarts.

Since version 17, Keycloak claims that it's capability for
"cloud native" deployments on Kubernetes has improved.

[Keycloak is offering a REST API](https://www.keycloak.org/docs-api/20.0.1/rest-api/index.html)
for all aspects of its administration interface.

For storage of Keycloak configuration and local user metadata
(e.g. from which external IdP a user account originally came from)
Keycloak supports several SQL backends through JDBC. Thus
it can be hooked up to a Postgres Database or to a
MariaDB/Galera cluster e.g..

### Zitadel

> Zitadel is a newer implementation of a SSO IdP. It is implemented
> in Go and under active development.
> 

Reviewer 3: ```suggestion
Zitadel is a newer implementation of a SSO IdP.
It is implemented in Go and under active development and maintained by ZITADEL.

The project is open for community [contributions](https://github.com/zitadel/zitadel/blob/main/CONTRIBUTING.md)  to all parts of the eco system.
Feature requests and bugs being tracked on [Github](https://github.com/orgs/zitadel/projects/2/views/5) for development.
Community questions can be asked in the [public chat](https://zitadel.com/chat) or via [Github Discussions](https://github.com/zitadel/zitadel/discussions). 

ZITADEL offers support for the commonly used authentication and authorization protocols such as OIDC, OAuth2, SAML2. 
It is a compliant and certified OpenID Connect provider with support for various Grant Types for both human users and machine users.

Compared to Keycloak SPIs, ZITADEL offers Actions to customize and integrate (eg, calling external APIs, Webhooks, customizing pre-built workflows, customizing tokens)
Actions are executed at runtime and can be maintained independently of platform.

Identity brokering (OIDC, SAML, JWT) can be configured system-wide or for each organization with templates.
Users will be created just in time for audit purposes and linked to the external identity provider.
Users can have multiple identity providers linked to their profile.
```

Reviewer 3: Just wanted to add a bit more substance to this section compared to keycloak. Feel free to adjust if you think it is too opinionated.

Author: Thanks for the input. I've picked it up.

It came to attention of the SCS project because it offers a
fresh take of an organization focussed data model, which has
the potential to simplify IdP federation to SCS customer domains
in the following areas:

> * For client services (single set of HTTP API endpoints).
> * For SCS operators for provisioning customer [organizations](https://zitadel.com/docs/concepts/structure/organizations)
>   and robust configuraton by using templated client, role and mapping
>   configuration.
> * For SCS customers for a robust user experience for self servicing.
> 

Reviewer 3: ```suggestion
* For client services (single set of HTTP API endpoints).
* For SCS operators for provisioning customer [organizations](https://zitadel.com/docs/concepts/structure/organizations)
  and robust configuraton by using templated client, role and mapping
  configuration.
* For SCS customers for a robust user experience for self servicing.

[Delegated Access Management](https://zitadel.com/docs/concepts/structure/organizations) reduces the management overhead compared to isolated realms.
Projects (Applications + Roles) can be maintained by one organization and delegated to be used by many other Organizations. 
Managers that receive granted Projects can assign users permissions to use the project.
```

Author: Thanks for the input, I've picked it up with minor adjustment in wording.

[Zitadel is offering REST APIs](https://zitadel.com/docs/apis/introduction)
for multiple areas of use and configuration.

In comparison to Keycloak it currently lacks support for the
[Device Authorization Grant](https://github.com/SovereignCloudStack/issues/issues/221),
which, at time of writing, is a feauture that is relevant
for SCS to be able use OpenStack CLI and APIs with federated
identities. At Zitadel there are at least two [Github issues](https://github.com/zitadel/oidc/issues/141)
proposed to implement this feature.

> At the time of writing it lacks support to consume identity data from LDAP backends.

Reviewer 3: Currently in progress with target end Q2. Glad for any feedback and use cases that we might not have considered:
https://github.com/zitadel/zitadel/issues/4609

Author: Thanks for the heads up, I included the link and adjusted the wording accordingly.

But that is currently not a relevant factor in the SCS context.

> 
> At time of writing an PoC "spike" is done to assess and verify the hopes

Reviewer 3: ```suggestion
ZITADEL can be deployed on [multiple platforms](https://zitadel.com/docs/self-hosting/deploy/overview).
Supported databases are CockroachDB and PostgreSQL.
For [production setups](https://zitadel.com/docs/self-hosting/manage/production) it is recommended to use Kubernetes (or similar like Knative) and CockroachDB.

At time of writing an PoC "spike" is done to assess and verify the hopes
```

Author: Included except for the first point which doesn't seem relevant to me here.

connected with Zitadel in the context of the SCS testbed.

# Open questions

* How would we implement testbed deployment support for Zitadel?
    * e.g. `wsgi-keystone.conf` would need to look different. One template covering both options?
    * e.g. steps like `openstack federation protocol create` would probably be different.
* Should we support both as options?
    * What's the benefit?
    * How would we allow SCS operators to choose?
* Do we need some kind of SWOT analysis to come to a decision?

# Decision

_Decision_

> 

Reviewer 2: As this document gathers the requirements, I’d remove the Decisions part and add it in a subsequent version of this document, if we actually took decisions.
